torch/distributed/_spmd/data_parallel.py

```diff
@@ -673,36 +707,39 @@ def mark_data_parallel_shardings(
             raise RuntimeError(f"op code {node.op} not supported")
 
 
-def _to_local_shard(
-    full_tensor: torch.Tensor, spec: DTensorSpec, from_local: bool = False
-) -> torch.Tensor:
+def _partition_val(val: Any, spec: DTensorSpec) -> Any:
     """
-    util function to convert a full tensor to its local component
+    util function to convert a full tensor val to its local component
     """
-    local_shard = full_tensor
-    if from_local:
-        # flag indicate the tensor is already local, we don't need to do anything
+    if isinstance(val, torch.Tensor):
+        local_shard = val
+        if val.ndim == 0:
+            # If it's already a scalar tensor, it is already local, we don't
+            # need to do anything
+            return local_shard
+
+        for idx, placement in enumerate(spec.placements):
+            if placement.is_shard():
```

Comment:
does this work if the tensor is sharded across a 2D mesh?

Reply:
Yes, it should work automatically as we are looping over the placements